network/builder/src/builder.rs

```diff
@@ -371,21 +376,39 @@ impl NetworkBuilder {
 
     pub fn add_connection_monitoring(&mut self) -> &mut Self {
         // Initialize and start HealthChecker.
-        let (hc_network_tx, hc_network_rx) = self.add_protocol_handler(health_checker::network_endpoint_config());
+        let (hc_network_tx, hc_network_rx) =
+            self.add_protocol_handler(health_checker::network_endpoint_config());
         let health_checker_config = HealthCheckerBuilderConfig::new(
             self.network_context.clone(),
             self.ping_interval_ms,
             self.ping_timeout_ms,
             self.ping_failures_tolerated,
-        );
-        let health_checker = health_checker::build_health_checker_from_config(
-            &self.executor,
-            health_checker_config,
             hc_network_tx,
             hc_network_rx,
         );
-        self.executor.spawn(health_checker.start());
-        debug!("{} Started health checker", self.network_context);
+
+        self.health_checker_cfg = Some(health_checker_config);
+
+        self.build_connection_monitoring()
+    }
+
+    fn build_connection_monitoring(&mut self) -> &mut Self {
+        if let Some(health_checker_config) = self.health_checker_cfg.take() {
```

Comment:
why do we have this indirection?

Reply:
Haha, "Outdated" do we fix this in a later PR?

Reply:
Looking forward $k$ steps, the goal is to have the following in network builder:

```
pub fn build(&mut self) {
  health_checker_builder.build();
  connectivity_monitor_builder.build();
  peer_manager_builder.build();
  ...
}

pub fn start(&mut self, executor) {
  health_checker.start(executor);
  connectivity_monitor.start(executor);
 ...
}
```
and main_node to have something like:

```
setup_environment() {
  for network in network_config {
     network_builders.push(network_builder::create(network));
  }
  consensus_builder = consensus.create(consensus_config)
  ....

  network_builders.build();
  consensus_builder.build();
  state_sync_builder.build();

  ...

  state_sync_builder.start();
  network_builders.start();
   consensus_builder.start();


}

```

The idea being to decouple configure/build/start and isolate race conditions as much as possible.

Reply:
> why do we have this indirection?

If you're asking about the config / object gap.... that distinction will go away shortly (within the next ~3 commits)

Reply:
> Haha, "Outdated" do we fix this in a later PR?

I have no idea how "outdated" shows up. :)

Maybe it gets fixed in a later commit that's in the PR????

Reply:
outdated means you have a massive stack!

Reply:
> Haha, "Outdated" do we fix this in a later PR?

I think outdated shows up when I rebase and then force push.

Reply:
> outdated means you have a massive stack!

22 commits and counting.